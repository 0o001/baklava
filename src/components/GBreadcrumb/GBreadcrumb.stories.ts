import { storiesOf } from '@storybook/vue';
import { withKnobs, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GBreadcrumb from './';

storiesOf('GBreadcrumb', module)
  .addDecorator(withKnobs)
  .add('default', () => ({
    components: { GBreadcrumb },
    props: {
      divider: {
        default: text('Divider', '>'),
      },
    },
    template: `
      <div style="padding: 30px;">
        <GBreadcrumb :items="items" @click="onClick" :divider="divider">
        </GBreadcrumb>
      </div>`,
    computed: {
      items () {
        return [
          { text: 'Breadcrumb 1', href: '/a/b/c' },
          { text: 'Breadcrumb 2', href: '/a/b/c', disabled: true },
          { text: 'Breadcrumb 3', href: '/a/b/c' },
        ];
      },
    },
    methods: {
      onClick (item) {
        action('Click a link from breadcrumb')(item);
      },
    },
  }));
