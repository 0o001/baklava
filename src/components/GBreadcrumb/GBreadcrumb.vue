<template>
  <GBox
    class="g-breadcrumb"
    flex
    align="items-center"
  >
    <GBox
      flex
      align="items-center"
      class="item-wrapper"
      v-for="(item, index) in items"
      :key="index"
    >
      <a
        :href="item.href"
        v-if="!isLastItem(index)"
        @click.prevent="onClick(item)"
        :class="{ disabled: item.disabled, item: true }"
      >
        <GText
          color="dark-grey-500"
          variant="menu-item"
        >
          {{ item.text }}
        </GText>
      </a>
      <GText
        v-else
        class="item"
        color="light-grey-500"
        variant="menu-item"
      >
        {{ item.text }}
      </GText>
      <GText
        v-if="!isLastItem(index)"
        class="divider"
        color="dark-grey-500"
        v-html="divider"
      />
    </GBox>
  </GBox>
</template>

<script lang="ts">

import GBox from '../GBox/GBox.vue';
import GText from '../GText/GText.vue';

type BreadcrumbOptions = {
  text: string;
  href: string;
  disabled: boolean;
}

export default {
  name: 'GBreadcrumb',
  components: {
    GBox,
    GText,
  },
  props: {
    items: {
      type: Array,
      default: () => [],
    },
    divider: {
      type: String,
      default: '>',
    },
  },
  computed: {

  },
  methods: {
    onClick (item: BreadcrumbOptions) {
      if (item.disabled) {
        return;
      }

      this.$emit('click', item);
    },

    isLastItem (itemIndex: number) {
      return this.items.length - 1 === itemIndex;
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-breadcrumb {
    .item-wrapper:not(:last-child) .item {
      color: var(--dark-grey-500);
    }

    .item-wrapper .item.disabled {
      cursor: default;
      pointer-events: none;
      text-decoration: none;
    }

    .divider {
      margin: 0 12px;
    }
  }

</style>
