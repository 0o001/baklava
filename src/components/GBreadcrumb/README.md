# GBreadcrumb

## Props

| Name      | Type     | Description | Default    |
| --------- | -------- | ----------- | ---------- |
| `items`   | `Array`  |             | `() => []` |
| `divider` | `String` |             | `">"`      |

## Events

| Name    | Description                                              |
| ------- | -------------------------------------------------------- |
| `click` | <br/>**Arguments**<br/><ul><li>**`item: any`**</li></ul> |

## Methods

### onClick()

**Syntax**

```typescript
onClick(item: BreadcrumbOptions): void
```

### isLastItem()

**Syntax**

```typescript
isLastItem(itemIndex: number): unknown
```

