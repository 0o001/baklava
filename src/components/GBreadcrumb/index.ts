import GBreadcrumb from './GBreadcrumb.vue';

export default GBreadcrumb;
