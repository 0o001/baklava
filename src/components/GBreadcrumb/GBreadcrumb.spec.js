import { shallowMount } from '@vue/test-utils';
import GBreadcrumb from './';

describe('GBreadcrumb', () => {
  let wrapper;
  const renderWrapper = (options) => {
    wrapper = shallowMount(GBreadcrumb, options);
  };
  const items = [
    { text: 'text 1', href: 'href 1' },
    { text: 'text 2', href: 'href 2', disabled: true },
    { text: 'text 3', href: 'href 3' },
  ];

  it('should match with snapshot when default props', () => {
    renderWrapper({
      propsData: {
        items,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render given divider', () => {
    const divider = '/';
    renderWrapper({ propsData: { items, divider } });

    expect(wrapper.find('.divider').exists()).toBeTruthy();
    expect(wrapper.find('.divider').text()).toBe(divider);
  });

  it('should emit the clicked link', async () => {
    renderWrapper({ propsData: { items } });

    const linkItem = wrapper.find('.item');

    linkItem.trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().click).toBeTruthy();
    expect(wrapper.emitted().click[0][0]).toEqual(items[0]);
  });

  it('should not emit the clicked link if item is disabled', async () => {
    renderWrapper({ propsData: { items } });

    const linkItem = wrapper.find('.item.disabled');

    linkItem.trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().click).toBeFalsy();
  });
});
