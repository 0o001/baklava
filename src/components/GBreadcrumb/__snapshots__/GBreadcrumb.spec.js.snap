// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GBreadcrumb should match with snapshot when default props 1`] = `
<gbox-stub
  align="items-center"
  class="g-breadcrumb"
  flex="true"
>
  <gbox-stub
    align="items-center"
    class="item-wrapper"
    flex="true"
  >
    <a
      class="item"
      href="href 1"
    >
      <gtext-stub
        color="dark-grey-500"
        variant="menu-item"
      >
        
        text 1
      
      </gtext-stub>
    </a>
     
    <gtext-stub
      class="divider"
      color="dark-grey-500"
      variant="body"
    >
      &gt;
    </gtext-stub>
  </gbox-stub>
  <gbox-stub
    align="items-center"
    class="item-wrapper"
    flex="true"
  >
    <a
      class="disabled item"
      href="href 2"
    >
      <gtext-stub
        color="dark-grey-500"
        variant="menu-item"
      >
        
        text 2
      
      </gtext-stub>
    </a>
     
    <gtext-stub
      class="divider"
      color="dark-grey-500"
      variant="body"
    >
      &gt;
    </gtext-stub>
  </gbox-stub>
  <gbox-stub
    align="items-center"
    class="item-wrapper"
    flex="true"
  >
    <gtext-stub
      class="item"
      color="light-grey-500"
      variant="menu-item"
    >
      
      text 3
    
    </gtext-stub>
     
    <!---->
  </gbox-stub>
</gbox-stub>
`;
